File: src/main/java/org/timmc/johnny/IPv4Host.java
package org.timmc.johnny;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Host component of a URI indicating an IPv4 address, represented as 4 integers.
 */
public class IPv4Host extends Host {
    /** See {@link #getRaw()} */
    public final String raw;
    /**
     * Get the 4 numeric segments of the IPv4 address
     */
    public final List<Integer> quad;

    /**
     * Create an IPv4Host from 4 integers. Raw form is derived.
     */
    public IPv4Host(int seg0, int seg1, int seg2, int seg3) {
        this(Arrays.asList(seg0, seg1, seg2, seg3));
    }

    /**
     * Create an IPv4Host from an array of 4 integers. Raw form is derived.
     */
    public IPv4Host(List<Integer> quad) {
        this(quad, IPv4Host.format(quad));
    }

    /**
     * Create an IPv4Host from an array of 4 integers and an accompanying
     * original raw form.
     * @param quad Array of 4 ints from 0 to 255 inclusive. Non-null.
     * @param raw Raw form of host before it was parsed. Non-null.
     */
    public IPv4Host(List<Integer> quad, String raw) {
        if (quad == null) {
            throw new NullPointerException("IPv4 quad must not be null");
        }
        // Copy, and then validate the copy
        this.quad = Collections.unmodifiableList(new ArrayList<>(quad));
        if (this.quad.size() != 4) {
            throw new IllegalArgumentException("IPv4 quad must contain 4 entries");
        }
        for (Integer seg: this.quad) {
            if (seg == null) {
                throw new NullPointerException("IPv4 segments must not be null");
            } else if (seg < 0) {
                throw new IllegalArgumentException("IPv4 segments must not be less than zero");
            } else if (seg > 255) {
                throw new IllegalArgumentException("IPv4 segments must not be greater than 255");
            }
        }
        if (raw == null) {
            throw new NullPointerException("IPv4 raw form must not be null");
        }
        this.raw = raw;
    }

    public String format() {
        return IPv4Host.format(quad);
    }

    /**
     * Format an IP v4 quad as a string, for URL representation.
     */
    public static String format(List<Integer> quad) {
        return quad.get(0) + "." + quad.get(1) + "." + quad.get(2) + "." + quad.get(3);
    }

    @Override
    public String getRaw() {
        return raw;
    }
}

File: src/main/java/org/timmc/johnny/IPv4Host.kt
package org.timmc.johnny

import java.lang.IllegalArgumentException

/**
 * Host component of a URI indicating an IPv4 address, represented as 4 integers.
 */
data class IPv4Host @JvmOverloads constructor (
    /** First segment of the address. */
    val seg0: Int,
    /** Second segment of the address. */
    val seg1: Int,
    /** Third segment of the address. */
    val seg2: Int,
    /** Fourth segment of the address. */
    val seg3: Int,
    override val raw: String = IPv4Host.format(seg0, seg1, seg2, seg3)
): Host() {
    init {
        if (seg0 !in 0..255 || seg1 !in 0..255 || seg2 !in 0..255 || seg3 !in 0..255)
            throw IllegalArgumentException("IPv4 segments must be numbers from 0 to 255, inclusive")
    }

    /** Create an IPv4Host from a list of 4 integers. */
    @JvmOverloads
    constructor(segments: List<Int>, raw: String = IPv4Host.format(segments))
        : this(segments[0], segments[1], segments[2], segments[3], raw)
    {
        if (segments.size != 4)
            throw IllegalArgumentException("IPv4 address cannot be constructed from list with more or fewer than 4 integers")
    }

    /**
     * Get the 4 numeric segments of the IPv4 address
     */
    val quad: List<Int> = listOf(seg0, seg1, seg2, seg3)

    override fun format(): String {
        return format(seg0, seg1, seg2, seg3)
    }

    companion object {
        private fun format(segments: List<Int>): String {
            return format(segments[0], segments[1], segments[2], segments[3])
        }

        private fun format(seg0: Int, seg1: Int, seg2: Int, seg3: Int): String {
            return "$seg0.$seg1.$seg2.$seg3"
        }
    }
}
